siobrultech_protocols/gem/api.py

```diff
@@ -92,9 +40,48 @@ async def send(arg: T) -> R:
         protocol.end_api_request()
 
 
+class GEMAPIStringResponseParser(Generic[R]):
+    """Most API responses in the GEM API end with \r\n; this parser waits until it sees a \r\n before attempting to parse."""
+
+    def __init__(self, parser: Callable[[str], R]) -> None:
+        self._parser = parser
+
+    def __call__(self, arg: str) -> R | None:
+        if not arg.endswith("\r\n"):
+            return None
+
+        return self._parser(arg)
+
+
+def parse_ecm_serial_number_from_settings(binary: bytes) -> int | None:
+    """
+    Unlike GEM, ECM-1240 doesn't have a specific get-serial-number API. Instead,
+    it returns the serial number as part of the device settings response.  This
+    parser understands enough of that format to extract the serial number from
+    the device settings.
+    """
+    if len(binary) < 33:
+        return None
+
+    # Note: The docs say that the serial number is transmitted big-endian, but
+    # my test device is outputting little-endian
+    [device_id, serial_number, zero, checksum] = struct.unpack_from(
```

Comment:
Can we check in their forum if the docs are wrong here of if there's a change of behavior?

Also, nit: punctuation at the end please

Reply:
Yeah, that's a good call. As I look at the docs more I see that the packet format docs are self-contradictory on this point, and I actually think that our current packet parsing is wrong. I sent Ben an email rather than posting in the forum because there's info in the question that probably shouldn't be in the forum and I had a thread going with him about this ECM-1240 support already. We'll see what he says.

Reply:
OK Ben pointed me to the implementation in https://github.com/BenK22/mtools/blob/master/bin/btcfg.py . It looks like indeed the serial in the packet is little-endian (#127 fixes our parsing of it) and the serial in the settings struct is big-endian. This PR is now ready to go.